File: Program1/Ruby/src/FixedPointList.rb
require './FixedPointNumber'
class FixedPointList
  attr_accessor :nums, :current_q

  def initialize
    super
    @nums = []
    @current_q = 12
  end
  def set_q(new_q_value)
    @current_q = new_q_value
  end


  def add(fixed_point_num)
    @nums.push(fixed_point_num)
  end

  def delete(fixed_point_num)
    (0...@nums.size).each do |i|
      if(@nums[i] == fixed_point_num)
        @nums.delete_at(i)
        return true
      end
    end
     false
  end

  def sum_all
    summed = FixedPointNumber.new(0,@current_q)
    @nums.each { |num|
      summed = summed.plus(num, @current_q)
    }
    return summed
  end

  def to_s
    if @nums.empty?
      return ""
    end

    out = @nums.join("\n")
    return out
  end
end

File: Program1/Java/src/FixedPointList.java
import java.util.ArrayList;
import java.util.Scanner;

public class FixedPointList {

    private ArrayList<FixedPointNumber> fixedPointList = new ArrayList<>();
    private int currentQ = 12;

    public int getCurrentQValue(){
        return currentQ;
    }

    public void setCurrentQValue(int q){
        this.currentQ = q;
    }

    public void add(FixedPointNumber p){
        fixedPointList.add(p);
    }

    public boolean delete(FixedPointNumber p){
        for (int i = 0; i < fixedPointList.size(); i++) {
            FixedPointNumber currentNum = fixedPointList.get(i);
            if(p.equals(currentNum)){
                fixedPointList.remove(currentNum);
                return true;
            }
        }
        return false;
    }

    public void print(){

        // While this is available for use due to requirements, It is not used because the current implementation
        // was more convenient for testing. That testing strategy defined how i implemented the
        // rest of program, so it is never used or needed.

        System.out.println("All fixed-point numbers in the list are:");
        System.out.println(this.toString());

    }

    public String toString(){
        StringBuilder builder = new StringBuilder();
        fixedPointList.forEach((num)-> builder.append('\n').append(num.toString()));
        return builder.toString();
    }

    public FixedPointNumber sumAll(){
        FixedPointNumber summed = new FixedPointNumber(0,currentQ);
        for (FixedPointNumber num : this.fixedPointList) {
            summed = summed.plus(num, this.currentQ);
        }
        return summed;
    }

}
